luminoso_api/v5_cli.py

```diff
@@ -68,10 +69,67 @@ def _read_params(input_file, json_body, p_params):
     return params
 
 
+def connect_with_token_args(args):
+    """
+    A shared function for working with Luminoso auth tokens in command line
+    utilities, interactively if necessary.
+    """
+    token = args.token
+    save_token = args.save_token
+    token_error = None
+
+    try:
+        client = LuminosoClient.connect(args.base_url, token=token)
+        # Make an authenticated request to test that the token is valid
+        client.get('/projects')
+    except LuminosoAuthError as e:
+        token_error = e
+
+    if token_error:
+        if token is not None:
+            # We were explicitly given a token that doesn't work, so re-raise
+            # the error
+            raise token_error
+        else:
+            url_parsed = urlparse(args.base_url)
+            print(
+                "You should be able to obtain an API token by going to "
+                "{scheme}://{netloc}/user.html and clicking the "
+                "'API tokens' button.".format(
+                    scheme=url_parsed.scheme, netloc=url_parsed.netloc
+                )
+            )
+            token = input('Enter your Luminoso API token: ')
+            if not token:
+                print("Cancelling because no token is available.")
+                raise SystemExit
+            if not save_token:
+                save_response = input(
+                    'Save this token to ~/.luminoso/tokens.json for future '
+                    'use? [Y/n] '
+                )
+                if save_response in ('', 'y', 'Y'):
+                    save_token = True
+                else:
+                    print('Not saving the token.')
```

Comment:
They will find out the token is invalid when the script proceeds with whatever it's doing, and saving an invalid token causes no harm on future runs of the script, because of the logic above which just asks the user for a valid token if they have an invalid one.

Is it really worth making another call to `/projects` to avoid saving an invalid token?

Reply:
I mostly think that this whole interface feels very clunky and like it's attempting to do the work of an exception, but manually.  The fact that it's also inconsistent is just another mark against it.